test/inductor/test_compiled_autograd.py

```diff
@@ -376,6 +376,170 @@ def eager_check():
+
+    def test_custom_fn_saved_shape_tensor(self):
+        def fn():
+            class MyFn(torch.autograd.Function):
+                @staticmethod
+                def forward(ctx, x):
+                    ctx.save_for_backward(x)
+                    return x
+
+                @staticmethod
+                def backward(ctx, gO):
+                    (x,) = ctx.saved_tensors
+                    return gO * x.shape[0]
+
+            for i in [10, 100, 10]:
+                x = torch.arange(0.0, i, requires_grad=True)
+                out = MyFn.apply(x)
+                loss = out.sum()
+                loss.backward()
+                yield x.grad
+
+        self.check_output_and_recompiles(fn, 2)
+
+    def test_custom_fn_saved_attr(self):
+        def fn():
+            class MyFn(torch.autograd.Function):
+                @staticmethod
+                def forward(ctx, x):
+                    ctx.shape = x.shape
+                    return x
+
+                @staticmethod
+                def backward(ctx, gO):
+                    x_shape = ctx.shape[0]
+                    return gO * x_shape
+
+            for i in [10, 100, 10]:
+                x = torch.arange(0.0, i, requires_grad=True)
+                out = MyFn.apply(x)
+                loss = out.sum()
+                loss.backward()
+                yield x.grad
+
+        with self.assertRaisesRegex(AssertionError, "Tensor-likes are not close!"):
+            self.check_output_and_recompiles(fn, 2)
+
+    def test_custom_fn_multiple_grads(self):
+        def fn():
+            class MyFn(torch.autograd.Function):
+                @staticmethod
+                def forward(ctx, x, y):
+                    return x + y
+
+                @staticmethod
+                def backward(ctx, gO):
+                    return gO, gO
+
+            for i in [10, 100, 10]:
+                x = torch.arange(0.0, i, requires_grad=True)
+                y = torch.arange(0.0, i, requires_grad=True)
+                out = MyFn.apply(x, y)
+                loss = out.sum()
+                loss.backward()
+                yield x.grad
+                yield y.grad
+
+        self.check_output_and_recompiles(fn, 2)
+
+    def test_custom_fns_with_same_graph(self):
+        def fn():
+            class MyFn1(torch.autograd.Function):
+                @staticmethod
+                def forward(ctx, x):
+                    return x
+
+                @staticmethod
+                def backward(ctx, gO):
+                    return gO
+
+            # same as MyFn1, but different autograd function id
+            # should not be using same graph as MyFn1
+            class MyFn2(torch.autograd.Function):
+                @staticmethod
+                def forward(ctx, x):
+                    return x
+
+                @staticmethod
+                def backward(ctx, gO):
+                    return gO
+
+            for myfn in [MyFn1, MyFn2]:
```

Comment:
```suggestion
            for myfn in [MyFn1, MyFn2, MyFn1, MyFn2]:
```
should still result in 2 recompiles, but also tests reuse